engine/classes/Elgg/Application.php

```diff
@@ -0,0 +1,378 @@
+
+		// @todo move loading plugins into a single boot function that replaces 'boot', 'system' event
+		// and then move this code in there.
+		// This validates the view type - first opportunity to do it is after plugins load.
+		$viewtype = elgg_get_viewtype();
+		if (!elgg_is_registered_viewtype($viewtype)) {
+			elgg_set_viewtype('default');
+		}
+
+		// @todo deprecate as plugins can use 'init', 'system' event
+		$events->trigger('plugins_boot', 'system');
+
+		// Complete the boot process for both engine and plugins
+		$events->trigger('init', 'system');
+
+		$CONFIG->boot_complete = true;
+
+		// System loaded and ready
+		$events->trigger('ready', 'system');
+	}
+
+	/**
+	 * Rewrite rules for PHP cli webserver used for testing. Do not use on production sites
+	 * as normal web server replacement.
+	 *
+	 * You need to explicitly point to index.php in order for router to work properly:
+	 *
+	 * <code>php -S localhost:8888 index.php</code>
+	 *
+	 * @return bool True if Elgg's router will handle the file, false if PHP should serve it directly
+	 */
+	protected function runPhpWebServer() {
+		$urlPath = parse_url($_SERVER["REQUEST_URI"], PHP_URL_PATH);
+
+		if (preg_match('/^\/cache\/(.*)$/', $urlPath, $matches)) {
+			$_GET['request'] = $matches[1];
+			$handler = new CacheHandler($this);
+			$handler->handleRequest($_GET, $_SERVER);
+			exit;
+		}
+
+		if (preg_match('/^\/export\/([A-Za-z]+)\/([0-9]+)\/?$/', $urlPath, $matches)) {
+			$_GET['view'] = $matches[1];
+			$_GET['guid'] = $matches[2];
+			require "{$this->engine_dir}/handlers/export_handler.php";
+			exit;
+		}
+
+		if (preg_match('/^\/export\/([A-Za-z]+)\/([0-9]+)\/([A-Za-z]+)\/([A-Za-z0-9\_]+)\/$/', $urlPath, $matches)) {
+			$_GET['view'] = $matches[1];
+			$_GET['guid'] = $matches[2];
+			$_GET['type'] = $matches[3];
+			$_GET['idname'] = $matches[4];
+			require "{$this->engine_dir}/handlers/export_handler.php";
+			exit;
+		}
+
+		if (preg_match("/^\/rewrite.php$/", $urlPath, $matches)) {
+			require "{$this->install_dir}/install.php";
+			exit;
+		}
+
+		if ($urlPath !== '/' && file_exists($this->install_dir . $urlPath)) {
+			// serve the requested resource as-is.
+			return false;
+		}
+
+		$_GET['__elgg_uri'] = $urlPath;
+		return true;
+	}
+
+	/**
+	 * Bootstraps core, plugins and handles the routing.
+	 *
+	 * @return bool False if Elgg wants the PHP CLI server to handle the request
+	 */
+	function run() {
+		if (php_sapi_name() === 'cli-server') {
+			if (!$this->runPhpWebServer()) {
+				// PHP will serve this file directly
+				return false;
+			}
+		}
+
+		$this->bootCore();
+
+		$router = _elgg_services()->router;
+		$request = _elgg_services()->request;
+
+		if (!$router->route($request)) {
+			forward('', '404');
+		}
+		return true;
+	}
+
+	/**
+	 * Get a singleton instance for testing
+	 *
+	 * @return \Elgg\Application
+	 * @access private
+	 * @internal temporary solution for unit testing
+	 */
+	public static function getTestInstance() {
```

Comment:
These were added a bit hastily to get phpunits working. There are a few tests that need access to the application that's created in the bootstrap. There's probably a better way